### PR TITLE
test: error throw for incompatible function parameter type.

### PR DESCRIPTION
**Pull request by jinangshah21**
Here I have excluded present() (lfortran_intrinsic_builtin function) because it is only expecting argument as integer but by definition we can pass different types in present.

Follow up to this some more issues I Identified is : 
1) We can have implicit casting for arguments
2) If array is passed then we have to check their shape as well.

### CODE REVIEW

**Comment by certik** [comment]
Here is the error at the CI in SciPy:
```
+ ../src/bin/lfortran --fixed-form --implicit-typing --implicit-interface --implicit-argument-casting --show-asr --no-color ./scipy/integrate/dop/dop853.f
semantic error: Type mismatch in argument at argument (2); passed `(integer, real, real[:], real[:], real[:], integer[:], return_type: void)` to `real`.
   --> ./scipy/integrate/dop/dop853.f:564:24 - 565:59
    |
564 |          IF (H.EQ.0.D0) H=HINIT853(N,FCN,X,Y,XEND,POSNEG,K1,K2,K3,IORD,
    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
...
    |
565 |         &                       HMAX,ATOL,RTOL,ITOL,RPAR,IPAR)
    | ...^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ 


Note: Please report unclear, confusing or incorrect messages as bugs at
```
The error is not super clear --- is it using a tuple? So some kind of an implied do loop printing?

It would be good to create an MRE for this one and decide how to treat it in this PR.

**Comment by gxyd** [comment]
You can use the below fortran program as MRE:

### MRE

```f90
SUBROUTINE sub(f)
    IMPLICIT DOUBLE PRECISION (A-H)
    EXTERNAL f
    H=HINIT853(f)
    RETURN
END
   
FUNCTION HINIT853(f)
    IMPLICIT DOUBLE PRECISION (A-H)
    RETURN
END 


```

### Output

```console
> lfortran --implicit-typing --implicit-interface --implicit-argument-casting --show-asr --no-color test.f90

semantic error: Type mismatch in argument at argument (1); passed `(return_type: real)` to `real`.
   --> test.f90:333:7
    |
333 |     H=HINIT853(f)
    |       ^^^^^^^^^^^ 


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

**Comment by gxyd** [comment]
I'll move the PR to draft for now, let us know @jinangshah21  if that helps.

**Comment by certik** [comment]
Thanks @gxyd for the MRE, very helpful!

**Comment by jinangshah21** [comment]
Here ```HINIT853(f)``` doesn't declare f as External function in its body so it is automatically considering argument type as 'real'.
If I am specifying ```EXTERNAL f```` in body of ```HINIT853(f)``` , code doesn't throw error.

In Gfortran : 

```PROGRAM test_present
  EXTERNAL func
  print *, f(func)
END PROGRAM

FUNCTION f(func)
    f = func(3)

real FUNCTION func(x)
    real::x
    f = x
  END FUNCTION```

is valid still we haven't declared func in f(). So I think we should exclude case when argument is a function and we are returning that funtion?

**Comment by Pranavchiku** [comment]
@jinangshah21 is this PR ready to review?

**Comment by jinangshah21** [comment]
This PR is checking Parameter types for all types except when Functions are passed in arguments because reason I have mentioned in previous comment.

**Comment by Pranavchiku** [comment]
I mean if you are done with making changes in this PR and I can review it and merge if seems fine? 😅

**Comment by jinangshah21** [comment]
Yes, I am done with this PR. You can review it.

**Comment by Pranavchiku** [comment]
Okay, please click `Ready for review` button:

image

**Review by Pranavchiku — APPROVED** [review]
looks good, just apply the change, update references and then this PR will be good to merge. Nice work, in next PR you can try to fix `present` so that we can remove special handling for it.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -4875,6 +4875,33 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
         }
         ASRUtils::set_absent_optional_arguments_to_null(args, func, al);
         legacy_array_sections_helper(v, args, loc);
+
+        ASR::FunctionType_t* func_type = ASRUtils::get_FunctionType(v);
+
+        // Currently present function is supporting only integer arguments
+        // After implementing present below if condition should be removed
```
Comment (Pranavchiku) [inline src/lfortran/semantics/ast_common_visitor.h:4882]: beautiful, I came to check if you have passed a comment and you have, perfect!

**Comment by gxyd** [comment]
I am not sure if we intend to add a test case for the MRE here: https://github.com/lfortran/lfortran/pull/4915#issuecomment-2376871626 in this PR or not. In case we don't intend to add it in this PR, we need to make sure to open an issue to add a test case, as it failed as a part of our third party, and not with our integration tests.

**Comment by certik** [comment]
> I am not sure if we intend to add a test case for the MRE here: https://github.com/lfortran/lfortran/pull/4915#issuecomment-2376871626 in this PR or not. In case we don't intend to add it in this PR, we need to make sure to open an issue to add a test case, as it failed as a part of our third party, and not with our integration tests.

How is it that previously it was failing and now it is not?

I would definitely add the test for the MRE that you found in this PR, to ensure things truly work.

**Comment by jinangshah21** [comment]
It is not failing as I excluded case when argument is a function because [comment](https://github.com/lfortran/lfortran/pull/4915#issuecomment-2377772871)

**Comment by Pranavchiku** [comment]
> How is it that previously it was failing and now it is not?

Now, he made a check to bypass `FunctionType` and hence it is passing.

**Comment by certik** [comment]
Let's add a test for for the other test case in this PR as suggested above?

**Comment by jinangshah21** [comment]
```
; ModuleID = 'LFortran'
source_filename = "LFortran"

define float @f(i32* %x) {
.entry:
  %f = alloca float, align 4
  br label %return

return:                                           ; preds = %.entry
  %0 = load float, float* %f, align 4
  ret float %0
}

define double @hinit853(double* %f) {
.entry:
  %hinit853 = alloca double, align 8
  br label %return

unreachable_after_return:                         ; No predecessors!
  br label %return

return:                                           ; preds = %unreachable_after_return, %.entry
  %0 = load double, double* %hinit853, align 8
  ret double %0
}

define void @sub(double ()* %f) {
.entry:
  %h = alloca double, align 8
  %0 = call double @hinit853(double ()* %f)
  store double %0, double* %h, align 8
  br label %return

unreachable_after_return:                         ; No predecessors!
  br label %return

return:                                           ; preds = %unreachable_after_return, %.entry
  ret void
}

declare double @f.1()

define i32 @main(i32 %0, i8** %1) {
.entry:
  call void @_lpython_call_initial_functions(i32 %0, i8** %1)
  call void @sub(float (i32*)* @f)
  br label %return

return:                                           ; preds = %.entry
  ret i32 0
}

declare void @_lpython_call_initial_functions(i32, i8**)
asr_to_llvm: module failed verification. Error:
Call parameter type does not match function signature!
double ()* %f
 double*  %0 = call double @hinit853(double ()* %f)
Call parameter type does not match function signature!
float (i32*)* @f
 double ()*  call void @sub(float (i32*)* @f)

code generation error: asr_to_llvm: module failed verification. Error:
Call parameter type does not match function signature!
double ()* %f
 double*  %0 = call double @hinit853(double ()* %f)
Call parameter type does not match function signature!
float (i32*)* @f
 double ()*  call void @sub(float (i32*)* @f)
```

This MRE is Failing with current main also and not affected by changes in this PR. So I think This should be a separate issue.

**Comment by jinangshah21** [comment]
```
; ModuleID = 'LFortran'
source_filename = "LFortran"

define double @hinit853(double* %f) {
.entry:
  %hinit853 = alloca double, align 8
  br label %return

unreachable_after_return:                         ; No predecessors!
  br label %return

return:                                           ; preds = %unreachable_after_return, %.entry
  %0 = load double, double* %hinit853, align 8
  ret double %0
}

define void @sub(double ()* %f) {
.entry:
  %h = alloca double, align 8
  %0 = call double @hinit853(double ()* %f)
  store double %0, double* %h, align 8
  br label %return

unreachable_after_return:                         ; No predecessors!
  br label %return

return:                                           ; preds = %unreachable_after_return, %.entry
  ret void
}

declare double @f()

define i32 @main(i32 %0, i8** %1) {
.entry:
  call void @_lpython_call_initial_functions(i32 %0, i8** %1)
  call void @sub(double ()* @f.1)
  br label %return

return:                                           ; preds = %.entry
  ret i32 0
}

declare double @f.1()

declare void @_lpython_call_initial_functions(i32, i8**)
asr_to_llvm: module failed verification. Error:
Call parameter type does not match function signature!
double ()* %f
 double*  %0 = call double @hinit853(double ()* %f)

code generation error: asr_to_llvm: module failed verification. Error:
Call parameter type does not match function signature!
double ()* %f
 double*  %0 = call double @hinit853(double ()* %f)
```
This MRE is Failing with current main also and not affected by changes in this PR. I think this should be a separate issue.

**Comment by Pranavchiku** [comment]
I checked it, and it generates ASR but not LLVM in both main and this PR with identical errors. OR I think we are missing a flag required to compile it with SciPy.

**Comment by Pranavchiku** [comment]
The CI failure was checking `--show-asr` for the corresponding file and the MRE now generates identical ASR with main. @jinangshah21 register it as a reference test and then this PR will be good to go.

**Comment by jinangshah21** [comment]
I have added MRE as a reference test.

File: tests/reference/asr-func_parameter_type_02-9155cd7.json [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{
+    "basename": "asr-func_parameter_type_02-9155cd7",
+    "cmd": "lfortran --show-asr --implicit-typing --implicit-interface --no-color {infile} -o {outfile}",
+    "infile": "tests/errors/func_parameter_type_02.f90",
```
Comment (certik) [inline tests/reference/asr-func_parameter_type_02-9155cd7.json:4]: This file does not give an error message, correct? So it should go into `tests` or even better into `integration_tests`.

Comment (certik) [reply tests/reference/asr-func_parameter_type_02-9155cd7.json:4]: This can be done in a subsequent PR.

**Review by certik — APPROVED** [review]
I think this is now good to merge, thanks!

File: tests/reference/asr-func_parameter_type_02-9155cd7.json (hunk 2 above)
Comment (Pranavchiku) [reply tests/reference/asr-func_parameter_type_02-9155cd7.json:4]: I opened https://github.com/lfortran/lfortran/issues/4948 for this.

**Pull request merged**